src/sse/sse_vector.rs

```diff
@@ -0,0 +1,231 @@
+// ```
+// let values = [
+//     output.store_complex(input[0], 0),
+//     output.store_complex(input[1], 1),
+//     output.store_complex(input[2], 2),
+//     output.store_complex(input[3], 3),
+// ];
+// ```
+macro_rules! write_complex_to_array {
+    ($input:ident, $output:ident, { $($idx:literal),* }) => {
+        $(
+            $output.store_complex($input[$idx], $idx);
+        )*
+    }
+}
+
+// Write the low half of these indexes of an array of simd vectors to the same indexes of an SseArray.
+// Takes a name of a vector to read from, one to write to, and a list of indexes.
+// This statement:
+// ```
+// let values = write_partial_lo_complex_to_array!(input, output, {0, 1, 2, 3});
+// ```
+// is equivalent to:
+// ```
+// let values = [
+//     output.store_partial_lo_complex(input[0], 0),
+//     output.store_partial_lo_complex(input[1], 1),
+//     output.store_partial_lo_complex(input[2], 2),
+//     output.store_partial_lo_complex(input[3], 3),
+// ];
+// ```
+macro_rules! write_partial_lo_complex_to_array {
+    ($input:ident, $output:ident, { $($idx:literal),* }) => {
+        $(
+            $output.store_partial_lo_complex($input[$idx], $idx);
+        )*
+    }
+}
+
+// Write these indexes of an array of simd vectors to the same indexes, multiplied by a stride, of an SseArray.
+// Takes a name of a vector to read from, one to write to, an integer stride, and a list of indexes.
+// This statement:
+// ```
+// let values = write_complex_to_array_separate!(input, output, {0, 1, 2, 3});
+// ```
+// is equivalent to:
+// ```
+// let values = [
+//     output.store_complex(input[0], 0),
+//     output.store_complex(input[1], 2),
+//     output.store_complex(input[2], 4),
+//     output.store_complex(input[3], 6),
+// ];
+// ```
+macro_rules! write_complex_to_array_strided {
+    ($input:ident, $output:ident, $stride:literal, { $($idx:literal),* }) => {
+        $(
+            $output.store_complex($input[$idx], $idx*$stride);
+        )*
+    }
+}
+
+pub trait SseArray {
+    type VectorType;
+    const COMPLEX_PER_VECTOR: usize;
+    unsafe fn load_complex(&self, index: usize) -> Self::VectorType;
+    unsafe fn load_partial1_complex(&self, index: usize) -> Self::VectorType;
+    unsafe fn load1_complex(&self, index: usize) -> Self::VectorType;
+}
+
+impl SseArray for RawSlice<Complex<f32>> {
+    type VectorType = __m128;
+    const COMPLEX_PER_VECTOR: usize = 2;
+
+    #[inline(always)]
+    unsafe fn load_complex(&self, index: usize) -> Self::VectorType {
+        debug_assert!(self.len() >= index + Self::COMPLEX_PER_VECTOR);
+        _mm_loadu_ps(self.as_ptr().add(index) as *const f32)
+    }
+
+    #[inline(always)]
+    unsafe fn load_partial1_complex(&self, index: usize) -> Self::VectorType {
+        debug_assert!(self.len() >= index + 1);
+        _mm_castpd_ps(_mm_load_sd(self.as_ptr().add(index) as *const f64))
+    }
+
+    #[inline(always)]
+    unsafe fn load1_complex(&self, index: usize) -> Self::VectorType {
```

Comment:
Do we need both load1_complex` and `load_partial1_complex`? If they're identical, maybe we could delete one.

Reply:
They are similar but not identical. The load1 copies the value to all elements of the vector, while load_partial1 zeroes the unused elements.